File: lib/plan-store.ts
import { mkdir, readdir, readFile, writeFile } from "fs/promises";
import path from "path";
import type { PlanStep } from "@/lib/plan-schema";

export type SavedPlan = {
  id: string;
  situation: string;
  createdAt: string;
  steps: PlanStep[];
};

const PLANS_DIR = path.join(process.cwd(), "data", "plans");

async function ensurePlansDir() {
  await mkdir(PLANS_DIR, { recursive: true });
}

export async function savePlan(record: SavedPlan): Promise<void> {
  await ensurePlansDir();
  const file = path.join(PLANS_DIR, `${record.id}.json`);
  await writeFile(file, JSON.stringify(record, null, 2), "utf8");
}

export async function getPlan(id: string): Promise<SavedPlan | null> {
  await ensurePlansDir();
  const safeId = path.basename(id);
  if (safeId !== id) return null;
  try {
    const raw = await readFile(path.join(PLANS_DIR, `${safeId}.json`), "utf8");
    return JSON.parse(raw) as SavedPlan;
  } catch {
    return null;
  }
}

export type PlanListItem = {
  id: string;
  situationPreview: string;
  createdAt: string;
};

export async function listPlansMeta(limit = 50): Promise<PlanListItem[]> {
  await ensurePlansDir();
  let names: string[];
  try {
    names = await readdir(PLANS_DIR);
  } catch {
    return [];
  }

  const jsonFiles = names.filter((n) => n.endsWith(".json"));
  const items: PlanListItem[] = [];

  for (const name of jsonFiles) {
    try {
      const raw = await readFile(path.join(PLANS_DIR, name), "utf8");
      const parsed = JSON.parse(raw) as Partial<SavedPlan>;
      if (
        typeof parsed.id === "string"
        && typeof parsed.situation === "string"
        && typeof parsed.createdAt === "string"
      ) {
        const preview =
          parsed.situation.length > 120
            ? `${parsed.situation.slice(0, 117)}...`
            : parsed.situation;
        items.push({
          id: parsed.id,
          situationPreview: preview,
          createdAt: parsed.createdAt,
        });
      }
    } catch {
      // skip corrupt entries
    }
  }

  items.sort((a, b) => (a.createdAt < b.createdAt ? 1 : -1));
  return items.slice(0, limit);
}

File: backend/app/services/plan_store.py
from __future__ import annotations

import json
from pathlib import Path

from app.config import PLANS_DIR
from app.models.plan import PlanListItem, SavedPlan, rollup_estimated_minutes


def _ensure_plans_dir() -> None:
    PLANS_DIR.mkdir(parents=True, exist_ok=True)


def save_plan(record: SavedPlan) -> None:
    _ensure_plans_dir()
    record = record.model_copy(
        update={"steps": rollup_estimated_minutes(record.steps)},
    )
    path = PLANS_DIR / f"{record.id}.json"
    path.write_text(
        json.dumps(record.model_dump(by_alias=True), indent=2),
        encoding="utf-8",
    )


def delete_plan(plan_id: str) -> bool:
    _ensure_plans_dir()
    safe_id = Path(plan_id).name
    if safe_id != plan_id:
        return False
    path = PLANS_DIR / f"{safe_id}.json"
    if not path.is_file():
        return False
    path.unlink()
    return True


def get_plan(plan_id: str) -> SavedPlan | None:
    _ensure_plans_dir()
    safe_id = Path(plan_id).name
    if safe_id != plan_id:
        return None
    path = PLANS_DIR / f"{safe_id}.json"
    if not path.is_file():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        plan = SavedPlan.model_validate(data)
        return plan.model_copy(
            update={"steps": rollup_estimated_minutes(plan.steps)},
        )
    except (json.JSONDecodeError, ValueError):
        return None


def list_plans_meta(limit: int = 50) -> list[PlanListItem]:
    _ensure_plans_dir()
    items: list[PlanListItem] = []
    try:
        names = list(PLANS_DIR.iterdir())
    except OSError:
        return []

    for path in names:
        if path.suffix != ".json":
            continue
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            pid = data.get("id")
            situation = data.get("situation")
            created = data.get("createdAt")
            if not (
                isinstance(pid, str)
                and isinstance(situation, str)
                and isinstance(created, str)
            ):
                continue
            preview = (
                f"{situation[:117]}..."
                if len(situation) > 120
                else situation
            )
            items.append(
                PlanListItem(
                    id=pid,
                    situationPreview=preview,
                    createdAt=created,
                )
            )
        except (json.JSONDecodeError, OSError, ValueError):
            continue

    items.sort(key=lambda x: x.created_at, reverse=True)
    return items[:limit]
